### ISSUE TITLE
Marking databases as read-only permanently

### ISSUE THREAD

**Issue by brainplot** [Kunzisoft/KeePassDX#1065]
Hi, I use my Android phone mainly as a read-only device for my KeePass database. I don't think it's ever happened so far that I've modified my database from my phone.

I like the fact KeePassDX allows you to toggle the Modifiable state of the database you're about to unlock. However, I dislike that I have to remember to toggle it each time if I want to be sure I won't accidentally modify my database. This is especially annoying since I use the fingerprint unlock and the fingerprint prompt immediately pops up when I select which database I want to open. That means that, in order to access the toggle, I have to first dismiss the fingerprint prompt, then toggle off the modifiable state and reactivate the fingerprint prompt by tapping on the fingerprint icon.

**Describe the solution you'd like**
I don't know what the ideal solution would be but I would personally consider either making it a general setting (something like "Open databases as read-only by default") or simply remembering what the last choice was for each file in the Recent Files list, with regards to the Modifiable toggle.

Thank you for KeePassDX. It's an amazing app!

**Comment by gabeweb** [comment]
I can't re-open my databases, only at the first time.

I thought it was an Android permission problem (I have had my databases stored in the internal memory of the phone), but even moving them to the SD memory the error persists.

![java.lang SecurityException: Permission Denial](https://user-images.githubusercontent.com/1930100/131277823-5a04c994-611b-4faa-ab2a-2df3be98ee0b.png)

`java.lang SecurityException: Permission Denial`

My phone: HONOR 10 Lite running Android 10 (EMUI 10.0).

**Comment by J-Jamet** [comment]
@brainplot Have you tried simply enabling the option in the settings? 
Enable -> Settings / App settings / Write-protected
Disable -> Settings / Advanced unlocking / Auto-open prompt

**Comment by J-Jamet** [comment]
@gabeweb Can you tell us what file manager you use?
Now KeePassDX rechecks the permissions granted by the file manager to see if the permissions are granted. If the file manager does not give them, it displays the error.

**Comment by gabeweb** [comment]
> @gabeweb Can you tell us what file manager you use?
> Now KeePassDX rechecks the permissions granted by the file manager to see if the permissions are granted. If the file manager does not give them, it displays the error.
> 
I'm using the default Android's file manager.


> What do you mean by "only the first time"?
When I open the database the first time in the app. But I've realized this: having the database and key file registered, I have to re-associate the key file without having to re-enter the database password, every time I have to open the database.

**Comment by brainplot** [comment]
> @brainplot Have you tried simply enabling the option in the settings? 
> Enable -> Settings / App settings / Write-protected
> Disable -> Settings / Advanced unlocking / Auto-open prompt

Shoot! I totally missed the option. I'm sorry! :cold_sweat:

Closing this.

**Comment by J-Jamet** [comment]
@brainplot No worries

@gabeweb Ok thanks, it is possible that the problem comes from the check that is not correctly done on the key file. I will try to improve this behavior.

I open the problem in a different issue.

**Comment by rmacklin** [comment]
@J-Jamet Are you open to this request from the original post:
> simply remembering what the last [write-protected] choice was for each file in the Recent Files list

?
I think it'd offer more flexibility than the current global setting - if you have one database that should be read-only and one that should be read-write, you would just have to open them in the desired state once and it would be remembered from that point on (unless you decide to change it, and then the new choice would be remembered).

**Comment by rmacklin** [comment]
@J-Jamet I see that you [thumbs-upped](https://github.com/user-attachments/assets/a4f40fd4-c7df-4da6-8e88-26e5034c0533) the comment, which I assumed means you'd accept a PR for that feature. I took a stab at it here:
- https://github.com/Kunzisoft/KeePassDX/pull/2100

I'm not really experienced in mobile development, but it seems to work as desired from my testing in the emulator. Not sure if the diff is how you'd have written the feature, though, so I'm open to any changes (and I also checked the "Allow edits by maintainers" box, so you're welcome to push directly to the branch).
